In [3]:
import pandas as pd


In [4]:
df = pd.read_csv("../data/AmesHousing.csv")


In [10]:
basic_df = pd.DataFrame({
    "A": [1, 2, 3],
    "B": [4, 5, 6]
})

# Any / All

any([True, False, False]) → is there at least one true?

all([True, True, True]) → is all of them true?

## Filtering Tricks

instead of repeating which looks bad like : `df.isnull().sum()[df.isnull().sum() > 0]`

In [ ]:
null_counts = df.isnull().sum()
null_counts[null_counts > 0]

## Adding new columns from existing df to other one


In [ ]:
df = pd.concat([df,encoded],axis=1)

# Creating Bins /Quantile Bins
e.g., can be used later for group by, creating histogram categories

bins =  **pd.cut()** = the distance-size of each group is same

quantile bins = **pd.qcut()** =  EVERY BIN HAS SAME NUMBER OF ELEMENTS


In [1]:
df["LotAreaBin"] = pd.qcut(df["Lot Area"], q=10)
#after creating bins
df.groupby("LotAreaBin")["Lot Frontage"].median()


NameError: name 'pd' is not defined

# Lambda, Apply, Map, Group By + Transform

In [ ]:
# remembering labda, they are botch identical:

lambda x : x + 2

def function(x):
    return x * 2

In [6]:
df['Lot Area'].head(5)

0    31770
1    11622
2    14267
3    11160
4    13830
Name: Lot Area, dtype: int64

## apply

* in series it just applies to elements
* in dataframes with axis = you can choose whether do you wanna apply columns or rows


In [9]:
# in series it just applies to elements
result = df['Lot Area'].apply(lambda x : x * 2)
result



0       63540
1       23244
2       28534
3       22320
4       27660
        ...  
2925    15874
2926    17770
2927    20882
2928    20020
2929    19254
Name: Lot Area, Length: 2930, dtype: int64

## apply vs map
With map(), the function receives one cell at a time — it has no awareness of the row or column it came from:

In [11]:
basic_df = pd.DataFrame({
    "A": [1, 2, 3],
    "B": [4, 5, 6]
})

result = basic_df.map(lambda x: x * 2)
# x = 1, then 4, then 2, then 5, then 3, then 6
# completely isolated scalar each time
result

,A,B
0,2,8
1,4,10
2,6,12


With apply(axis=1), the function receives an entire row as a Series:

In [13]:
result = basic_df.apply(lambda x: x["A"] + x["B"], axis=1)
# x = [1, 4], then [2, 5], then [3, 6]
# x is a full row, you index into it by column name
result

0    5
1    7
2    9
dtype: int64

## transform

In [16]:
medium_df = pd.DataFrame({
    "name":     ["Alice", "Bob", "Carol", "David", "Eve"],
    "category": ["electronics", "clothing", "electronics", "clothing", "electronics"],
    "price":    [999, 49, 1200, 89, 349],
    "units":    [3, 15, 1, 22, 8],
    "discount": [0.10, 0.05, 0.20, 0.05, 0.15]
})

In [20]:
medium_df

,name,category,price,units,discount
0,Alice,electronics,999,3,0.10
1,Bob,clothing,49,15,0.05
2,Carol,electronics,1200,1,0.20
3,David,clothing,89,22,0.05
4,Eve,electronics,349,8,0.15


In [18]:
result = medium_df.groupby("category")["price"].transform("mean")
result

0    849.333333
1     69.000000
2    849.333333
3     69.000000
4    849.333333
Name: price, dtype: float64

   849.333333   # electronics mean

   69.000000   # clothing mean

  849.333333   # electronics mean

  69.000000   # clothing mean
 849.333333   # electronics mean